python/pyarrow/tests/test_dataset.py

```diff
@@ -4591,6 +4591,22 @@ def file_visitor(written_file):
     assert result1.to_table().equals(result2.to_table())
 
 
+@pytest.mark.parquet
+@pytest.mark.pandas
+def test_write_dataset_use_threads_preserve_order(tempdir):
```

Comment:
Added test code from issue description. The test fails with `preserve_order=False`.